File: linked_list/SingleLL/SingleLL.c
// append: 리스트 끝에 노드 추가.
// prepend: 리스트 처음에 노드 추가.
// insertAfter: 특정 값 뒤에 노드 삽입.
// deleteNode: 특정 값을 가진 노드 삭제.
// search: 리스트에서 특정 값 검색.
// size: 리스트의 노드 개수 반환.
// reverse: 리스트를 뒤집음.
// printList: 리스트를 보기 좋게 출력.
// freeList: 메모리 해제.
// Sorting: Bubble Sort, Selection Sort, Insertion Sort, Merge Sort, Quick Sort

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Node 구조체 정의
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// 단일 연결 리스트 구조체 정의
typedef struct LinkedList {
    Node* head;
} LinkedList;

// 연결 리스트 초기화
LinkedList* createLinkedList() {
    LinkedList* list = (LinkedList*)malloc(sizeof(LinkedList));
    list->head = NULL;
    return list;
}

// 노드 생성
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// 리스트 끝에 노드 추가
void append(LinkedList* list, int data) {
    Node* newNode = createNode(data);
    if (list->head == NULL) {
        list->head = newNode;
        return;
    }
    Node* temp = list->head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = newNode;
}

// 리스트 처음에 노드 추가
void prepend(LinkedList* list, int data) {
    Node* newNode = createNode(data);
    newNode->next = list->head;
    list->head = newNode;
}

// 특정 값 뒤에 노드 삽입
void insertAfter(LinkedList* list, int prevData, int data) {
    Node* temp = list->head;
    while (temp != NULL && temp->data != prevData) {
        temp = temp->next;
    }
    if (temp == NULL) {
        printf("Error: Node with data %d not found.\n", prevData);
        return;
    }
    Node* newNode = createNode(data);
    newNode->next = temp->next;
    temp->next = newNode;
}

// 특정 값을 가진 노드 삭제
void deleteNode(LinkedList* list, int data) {
    if (list->head == NULL) {
        printf("Error: List is empty.\n");
        return;
    }
    if (list->head->data == data) {
        Node* temp = list->head;
        list->head = list->head->next;
        free(temp);
        return;
    }
    Node* temp = list->head;
    while (temp->next != NULL && temp->next->data != data) {
        temp = temp->next;
    }
    if (temp->next == NULL) {
        printf("Error: Node with data %d not found.\n", data);
        return;
    }
    Node* nodeToDelete = temp->next;
    temp->next = temp->next->next;
    free(nodeToDelete);
}

// 리스트에서 특정 값 검색
bool search(LinkedList* list, int data) {
    Node* temp = list->head;
    while (temp != NULL) {
        if (temp->data == data) {
            return true;
        }
        temp = temp->next;
    }
    return false;
}

// 리스트 크기 반환
int size(LinkedList* list) {
    int count = 0;
    Node* temp = list->head;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// 리스트 반전
void reverse(LinkedList* list) {
    Node* prev = NULL;
    Node* current = list->head;
    Node* next = NULL;
    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    list->head = prev;
}

// 리스트 출력
void printList(LinkedList* list) {
    Node* temp = list->head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

// 리스트 메모리 해제
void freeList(LinkedList* list) {
    Node* temp = list->head;
    while (temp != NULL) {
        Node* next = temp->next;
        free(temp);
        temp = next;
    }
    free(list);
}

// Bubble Sort
void bubbleSort(Node** headRef) {
    if (*headRef == NULL || (*headRef)->next == NULL)
        return;

    int swapped;
    Node* current;
    Node* end = NULL;

    do {
        swapped = 0;
        current = *headRef;

        while (current->next != end) {
            if (current->data > current->next->data) {
                int temp = current->data;
                current->data = current->next->data;
                current->next->data = temp;
                swapped = 1;
            }
            current = current->next;
        }
        end = current;
    } while (swapped);
}

// Selection Sort for Singly Linked List
void selectionSort(Node** headRef) {
    Node* current = *headRef;

    while (current != NULL) {
        Node* minNode = current;
        Node* nextNode = current->next;

        while (nextNode != NULL) {
            if (nextNode->data < minNode->data) {
                minNode = nextNode;
            }
            nextNode = nextNode->next;
        }

        // Swap data between current node and minNode
        if (minNode != current) {
            int temp = current->data;
            current->data = minNode->data;
            minNode->data = temp;
        }

        current = current->next;
    }
}

// Insertion Sort
void insertionSort(Node** headRef) {
    Node* sorted = NULL;
    Node* current = *headRef;

    while (current != NULL) {
        Node* next = current->next;

        if (sorted == NULL || sorted->data >= current->data) {
            current->next = sorted;
            sorted = current;
        } else {
            Node* temp = sorted;
            while (temp->next != NULL && temp->next->data < current->data) {
                temp = temp->next;
            }
            current->next = temp->next;
            temp->next = current;
        }

        current = next;
    }

    *headRef = sorted;
}

// Merge Sort Helper: Split the list into two halves
void splitList(Node* source, Node** frontRef, Node** backRef) {
    Node* slow = source;
    Node* fast = source->next;

    while (fast != NULL) {
        fast = fast->next;
        if (fast != NULL) {
            slow = slow->next;
            fast = fast->next;
        }
    }

    *frontRef = source;
    *backRef = slow->next;
    slow->next = NULL;
}

// Merge Sort Helper: Merge two sorted lists
Node* sortedMerge(Node* a, Node* b) {
    if (a == NULL)
        return b;
    if (b == NULL)
        return a;

    Node* result;
    if (a->data <= b->data) {
        result = a;
        result->next = sortedMerge(a->next, b);
    } else {
        result = b;
        result->next = sortedMerge(a, b->next);
    }
    return result;
}

// Merge Sort
void mergeSort(Node** headRef) {
    Node* head = *headRef;
    if (head == NULL || head->next == NULL)
        return;

    Node* a;
    Node* b;

    splitList(head, &a, &b);

    mergeSort(&a);
    mergeSort(&b);

    *headRef = sortedMerge(a, b);
}

// Helper function (for Quick Sort) to get the tail of the list
Node* getTail(Node* current) {
    while (current != NULL && current->next != NULL) {
        current = current->next;
    }
    return current;
}

// Partition function for Quick Sort
Node* partition(Node* head, Node* end, Node** newHead, Node** newEnd) {
    Node* pivot = end;
    Node* prev = NULL;
    Node* current = head;
    Node* tail = pivot;

    while (current != pivot) {
        if (current->data < pivot->data) {
            if (*newHead == NULL) {
                *newHead = current;
            }
            prev = current;
            current = current->next;
        } else {
            if (prev != NULL) {
                prev->next = current->next;
            }
            Node* temp = current->next;
            current->next = NULL;
            tail->next = current;
            tail = current;
            current = temp;
        }
    }

    if (*newHead == NULL) {
        *newHead = pivot;
    }
    *newEnd = tail;

    return pivot;
}

// Quick Sort recursive function
Node* quickSortRec(Node* head, Node* end) {
    if (head == NULL || head == end) {
        return head;
    }

    Node* newHead = NULL;
    Node* newEnd = NULL;

    Node* pivot = partition(head, end, &newHead, &newEnd);

    if (newHead != pivot) {
        Node* temp = newHead;
        while (temp->next != pivot) {
            temp = temp->next;
        }
        temp->next = NULL;

        newHead = quickSortRec(newHead, temp);

        temp = getTail(newHead);
        temp->next = pivot;
    }

    pivot->next = quickSortRec(pivot->next, newEnd);

    return newHead;
}

// Quick Sort
void quickSort(Node** headRef) {
    *headRef = quickSortRec(*headRef, getTail(*headRef));
}

int main() {
    // 연결 리스트 생성
    LinkedList* list = createLinkedList();

    // 1. append: 리스트 끝에 노드 추가
    append(list, 3);
    append(list, 1);
    append(list, 4);
    append(list, 2);
    printf("After append operations (3, 1, 4, 2):\n");
    printList(list);  // Output: 3 -> 1 -> 4 -> 2 -> NULL

    // 2. prepend: 리스트 처음에 노드 추가
    prepend(list, 0);
    printf("\nAfter prepend operation (0):\n");
    printList(list);  // Output: 0 -> 3 -> 1 -> 4 -> 2 -> NULL

    // 3. insertAfter: 특정 값 뒤에 노드 삽입
    insertAfter(list, 3, 5);
    printf("\nAfter inserting 5 after 3:\n");
    printList(list);  // Output: 0 -> 3 -> 5 -> 1 -> 4 -> 2 -> NULL

    // 4. deleteNode: 특정 값을 가진 노드 삭제
    deleteNode(list, 5);
    printf("\nAfter deleting 5:\n");
    printList(list);  // Output: 0 -> 3 -> 1 -> 4 -> 2 -> NULL

    // 5. search: 리스트에서 특정 값 검색
    printf("\nSearch for 4: %s\n", search(list, 4) ? "Found" : "Not Found");  // Output: Found
    printf("Search for 7: %s\n", search(list, 7) ? "Found" : "Not Found");    // Output: Not Found

    // 6. size: 리스트의 노드 개수 반환
    printf("\nSize of the list: %d\n", size(list));  // Output: 5

    // 7. reverse: 리스트를 뒤집음
    reverse(list);
    printf("\nAfter reversing the list:\n");
    printList(list);  // Output: 2 -> 4 -> 1 -> 3 -> 0 -> NULL

    // 8. Sorting: Bubble Sort
    bubbleSort(&list->head);
    printf("\nAfter Bubble Sort:\n");
    printList(list);  // Output: 0 -> 1 -> 2 -> 3 -> 4 -> NULL

    // 9. Sorting: Selection Sort
    append(list, 3);
    append(list, 1);
    printf("\nAfter adding 3 and 1 for Selection Sort:\n");
    printList(list);  // Output: 0 -> 1 -> 2 -> 3 -> 4 -> 3 -> 1 -> NULL

    selectionSort(&list->head);
    printf("\nAfter Selection Sort:\n");
    printList(list);  // Output: 0 -> 1 -> 1 -> 2 -> 3 -> 3 -> 4 -> NULL

    // 10. Sorting: Insertion Sort
    append(list, 5);
    append(list, 0);
    printf("\nAfter adding 5 and 0 for Insertion Sort:\n");
    printList(list);  // Output: 0 -> 1 -> 1 -> 2 -> 3 -> 3 -> 4 -> 5 -> 0 -> NULL

    insertionSort(&list->head);
    printf("\nAfter Insertion Sort:\n");
    printList(list);  // Output: 0 -> 0 -> 1 -> 1 -> 2 -> 3 -> 3 -> 4 -> 5 -> NULL

    // 11. Sorting: Merge Sort
    printf("\nAfter Merge Sort:\n");
    mergeSort(&list->head);
    printList(list);  // Output: 0 -> 0 -> 1 -> 1 -> 2 -> 3 -> 3 -> 4 -> 5 -> NULL

    // 12. Sorting: Quick Sort
    append(list, 6);
    append(list, 2);
    printf("\nAfter adding 6 and 2 for Quick Sort:\n");
    printList(list);  // Output: 0 -> 0 -> 1 -> 1 -> 2 -> 3 -> 3 -> 4 -> 5 -> 6 -> 2 -> NULL

    quickSort(&list->head);
    printf("\nAfter Quick Sort:\n");
    printList(list);  // Output: 0 -> 0 -> 1 -> 1 -> 2 -> 2 -> 3 -> 3 -> 4 -> 5 -> 6 -> NULL

    // 메모리 해제
    freeList(list);

    return 0;
}

File: linked_list/SingleLL/SingleLL.py
# append: 끝에 노드 추가.
# prepend: 처음에 노드 추가.
# insert_after: 특정 값 뒤에 노드 삽입.
# delete: 특정 값을 가진 노드 삭제.
# search: 특정 값을 가진 노드 검색.
# size: 리스트의 크기 반환.
# reverse: 리스트를 뒤집음.
# print_list: 리스트를 보기 좋게 출력.
# has_cycle: 리스트에 사이클이 있는지 확인.
# remove_duplicates: 중복 노드 제거.
# find_nth_to_last: 끝에서 n번째 노드 찾기.
# Sorting: Bubble Sort, Selection Sort, Insertion Sort, Merge Sort, Quick Sort

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    # 1. Append a node at the end
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    # 2. Prepend a node at the beginning
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    # 3. Insert a node after a specific node
    def insert_after(self, prev_data, data):
        current = self.head
        while current and current.data != prev_data:
            current = current.next
        if current is None:
            raise ValueError(f"Node with data {prev_data} not found.")
        new_node = Node(data)
        new_node.next = current.next
        current.next = new_node

    # 4. Delete a node by value
    def delete(self, data):
        if not self.head:
            return
        if self.head.data == data:
            self.head = self.head.next
            return
        current = self.head
        while current.next and current.next.data != data:
            current = current.next
        if current.next is None:
            raise ValueError(f"Node with data {data} not found.")
        current.next = current.next.next

    # 5. Search for a node by value
    def search(self, data):
        current = self.head
        while current:
            if current.data == data:
                return True
            current = current.next
        return False

    # 6. Get size of the linked list
    def size(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count

    # 7. Reverse the linked list
    def reverse(self):
        prev = None
        current = self.head
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        self.head = prev

    # 8. Print the linked list
    def print_list(self):
        current = self.head
        result = []
        while current:
            result.append(current.data)
            current = current.next
        print(" -> ".join(map(str, result)))

    # 9. Detect a cycle in the linked list
    def has_cycle(self):
        slow, fast = self.head, self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

    # 10. Remove duplicates
    def remove_duplicates(self):
        current = self.head
        seen = set()
        prev = None
        while current:
            if current.data in seen:
                prev.next = current.next
            else:
                seen.add(current.data)
                prev = current
            current = current.next

    # 11. Find the nth-to-last node
    def find_nth_to_last(self, n):
        slow = self.head
        fast = self.head
        for _ in range(n):
            if not fast:
                raise ValueError(f"List has fewer than {n} elements.")
            fast = fast.next
        while fast:
            slow = slow.next
            fast = fast.next
        return slow.data

    # 12. Merge two sorted linked lists (static method)
    @staticmethod
    def merge_sorted(l1, l2):
        dummy = Node(0)
        current = dummy
        while l1 and l2:
            if l1.data < l2.data:
                current.next = l1
                l1 = l1.next
            else:
                current.next = l2
                l2 = l2.next
            current = current.next
        current.next = l1 or l2
        return dummy.next
    
    # Bubble Sort
    def bubble_sort(self):
        if not self.head or not self.head.next:
            return

        end = None
        while end != self.head:
            current = self.head
            while current.next != end:
                next_node = current.next
                if current.data > next_node.data:
                    current.data, next_node.data = next_node.data, current.data
                current = current.next
            end = current

    # Selection Sort
    def selection_sort(self):
        current = self.head

        while current:
            min_node = current
            next_node = current.next

            while next_node:
                if next_node.data < min_node.data:
                    min_node = next_node
                next_node = next_node.next

            # Swap data
            if min_node != current:
                current.data, min_node.data = min_node.data, current.data

            current = current.next

    # Insertion Sort
    def insertion_sort(self):
        if not self.head or not self.head.next:
            return

        sorted_list = None
        current = self.head

        while current:
            next_node = current.next
            if not sorted_list or sorted_list.data >= current.data:
                current.next = sorted_list
                sorted_list = current
            else:
                temp = sorted_list
                while temp.next and temp.next.data < current.data:
                    temp = temp.next
                current.next = temp.next
                temp.next = current
            current = next_node

        self.head = sorted_list

    # Merge Sort
    def merge_sort(self):
        if not self.head or not self.head.next:
            return

        def split(head):
            slow = head
            fast = head.next
            while fast and fast.next:
                slow = slow.next
                fast = fast.next.next
            middle = slow.next
            slow.next = None
            return head, middle

        def merge(left, right):
            dummy = Node(0)
            tail = dummy
            while left and right:
                if left.data <= right.data:
                    tail.next = left
                    left = left.next
                else:
                    tail.next = right
                    right = right.next
                tail = tail.next
            tail.next = left or right
            return dummy.next

        def merge_sort_rec(head):
            if not head or not head.next:
                return head
            left, right = split(head)
            left = merge_sort_rec(left)
            right = merge_sort_rec(right)
            return merge(left, right)

        self.head = merge_sort_rec(self.head)

    # Quick Sort
    def quick_sort(self):
        def partition(start, end):
            pivot = start
            current = start.next
            tail = pivot

            while current != end:
                next_node = current.next
                if current.data < pivot.data:
                    # Move current to the front of the partition
                    tail.next = current.next
                    current.next = start
                    start = current
                else:
                    tail = current
                current = next_node

            return pivot, start

        def quick_sort_rec(start, end):
            if start == end or start.next == end:
                return start

            pivot, new_head = partition(start, end)
            pivot.next = quick_sort_rec(pivot.next, end)
            return quick_sort_rec(new_head, pivot)

        self.head = quick_sort_rec(self.head, None)

def main():
    # 새로운 SinglyLinkedList 생성
    ll = SinglyLinkedList()

    # 1. Append: 끝에 노드 추가
    ll.append(1)
    ll.append(2)
    ll.append(3)
    print("After append operations (1, 2, 3):")
    ll.print_list()  # Output: 1 -> 2 -> 3

    # 2. Prepend: 처음에 노드 추가
    ll.prepend(0)
    print("\nAfter prepend operation (0):")
    ll.print_list()  # Output: 0 -> 1 -> 2 -> 3

    # 3. Insert After: 특정 값 뒤에 노드 삽입
    ll.insert_after(2, 4)
    print("\nAfter inserting 4 after 2:")
    ll.print_list()  # Output: 0 -> 1 -> 2 -> 4 -> 3

    # 4. Delete: 특정 값을 가진 노드 삭제
    ll.delete(2)
    print("\nAfter deleting 2:")
    ll.print_list()  # Output: 0 -> 1 -> 4 -> 3

    # 5. Search: 특정 값을 가진 노드 검색
    print("\nSearch for 4:", ll.search(4))  # Output: True
    print("Search for 2:", ll.search(2))    # Output: False

    # 6. Size: 리스트의 크기 반환
    print("\nSize of the list:", ll.size())  # Output: 4

    # 7. Reverse: 리스트를 뒤집음
    ll.reverse()
    print("\nAfter reversing the list:")
    ll.print_list()  # Output: 3 -> 4 -> 1 -> 0

    # 8. Remove Duplicates: 중복 노드 제거
    ll.append(4)
    ll.append(3)
    print("\nAfter adding duplicates (4, 3):")
    ll.print_list()  # Output: 3 -> 4 -> 1 -> 0 -> 4 -> 3

    ll.remove_duplicates()
    print("\nAfter removing duplicates:")
    ll.print_list()  # Output: 3 -> 4 -> 1 -> 0

    # 9. Find Nth-to-Last Node: 끝에서 n번째 노드 찾기
    print("\n2nd-to-last node:", ll.find_nth_to_last(2))  # Output: 1

    # 10. Check for Cycle: 사이클 확인
    print("\nCheck if the list has a cycle:", ll.has_cycle())  # Output: False

    # 11. Sorting Examples
    # 재정렬을 위해 리스트 초기화
    ll = SinglyLinkedList()
    ll.append(3)
    ll.append(1)
    ll.append(4)
    ll.append(2)

    print("\nUnsorted list:")
    ll.print_list()  # Output: 3 -> 1 -> 4 -> 2

    # Bubble Sort
    ll.bubble_sort()
    print("\nAfter Bubble Sort:")
    ll.print_list()  # Output: 1 -> 2 -> 3 -> 4

    # 다시 정렬되지 않은 리스트로 초기화
    ll = SinglyLinkedList()
    ll.append(4)
    ll.append(3)
    ll.append(1)
    ll.append(2)

    # Selection Sort
    ll.selection_sort()
    print("\nAfter Selection Sort:")
    ll.print_list()  # Output: 1 -> 2 -> 3 -> 4

    # 다시 정렬되지 않은 리스트로 초기화
    ll = SinglyLinkedList()
    ll.append(3)
    ll.append(4)
    ll.append(1)
    ll.append(2)

    # Insertion Sort
    ll.insertion_sort()
    print("\nAfter Insertion Sort:")
    ll.print_list()  # Output: 1 -> 2 -> 3 -> 4

    # 다시 정렬되지 않은 리스트로 초기화
    ll = SinglyLinkedList()
    ll.append(4)
    ll.append(2)
    ll.append(1)
    ll.append(3)

    # Merge Sort
    ll.merge_sort()
    print("\nAfter Merge Sort:")
    ll.print_list()  # Output: 1 -> 2 -> 3 -> 4

    # 다시 정렬되지 않은 리스트로 초기화
    ll = SinglyLinkedList()
    ll.append(3)
    ll.append(1)
    ll.append(4)
    ll.append(2)

    # Quick Sort
    ll.quick_sort()
    print("\nAfter Quick Sort:")
    ll.print_list()  # Output: 1 -> 2 -> 3 -> 4

    # 12. Merge Two Sorted Lists
    print("\nMerging two sorted lists:")
    list1 = SinglyLinkedList()
    list1.append(1)
    list1.append(3)
    list1.append(5)

    list2 = SinglyLinkedList()
    list2.append(2)
    list2.append(4)
    list2.append(6)

    merged_head = SinglyLinkedList.merge_sorted(list1.head, list2.head)
    merged_list = SinglyLinkedList()
    merged_list.head = merged_head

    merged_list.print_list()  # Output: 1 -> 2 -> 3 -> 4 -> 5 -> 6


if __name__ == "__main__":
    main()